logic/src/androidMain/kotlin/com/wire/kalium/logic/sync/WorkSchedulerImpl.kt

```diff
@@ -90,43 +66,97 @@ internal actual class GlobalWorkSchedulerImpl(
+                .firstOrNull()
+                ?.firstOrNull()
+                ?.let {
+                    if (it.state == androidx.work.WorkInfo.State.ENQUEUED && it.runAttemptCount > 0) {
+                        // it means that the worker has been retried at least once, so update to reset the backoff
+                        WorkManager.getInstance(appContext).enqueueUniquePeriodicWork(
+                            UserConfigSyncWorker.NAME + userId.value + "-periodic",
+                            ExistingPeriodicWorkPolicy.UPDATE,
+                            buildConnectedPeriodicWorkRequest(UserConfigSyncWorker::class, userId, true)
+                        )
+                    }
+                }
+        }
     }
 }
+
+private fun buildConnectedPeriodicWorkRequest(
+    worker: KClass<out DefaultWorker>,
+    userId: UserId? = null,
+    resetBackoff: Boolean = false,
+): PeriodicWorkRequest {
+    val inputData = WrapperWorkerFactory.workData(worker, userId)
+    val connectedConstraint = Constraints.Builder()
+        .setRequiredNetworkType(NetworkType.CONNECTED)
+        .build()
+    val localTimeZone = TimeZone.currentSystemDefault()
+    val timeNow: Instant = DateTimeUtil.currentInstant() // current time
+    val timeScheduledToExecute = timeNow.toLocalDateTime(localTimeZone) // time at which the today's execution should take place
+        .let { localDateTimeNow ->
+            LocalDateTime(
+                localDateTimeNow.year, localDateTimeNow.monthNumber, localDateTimeNow.dayOfMonth,
+                TIME_OF_EXECUTION, 0, 0, 0
+            ).toInstant(localTimeZone)
+        }
+    val initialDelayMillis = // delay calculated as a difference between now and next scheduled execution
+        if (timeScheduledToExecute > timeNow) (timeScheduledToExecute - timeNow).inWholeMilliseconds
+        else (timeScheduledToExecute.plus(1, DateTimeUnit.DAY, localTimeZone) - timeNow).inWholeMilliseconds
+    return PeriodicWorkRequest.Builder(workerClass, REPEAT_INTERVAL, TimeUnit.HOURS)
+        .setInitialDelay(initialDelayMillis, TimeUnit.MILLISECONDS)
+        .setConstraints(connectedConstraint)
+        .setInputData(inputData)
+        .let { if (resetBackoff) it.setNextScheduleTimeOverride(System.currentTimeMillis()) else it.clearNextScheduleTimeOverride() }
+        .build()
+}
+
+private fun buildConnectedOneTimeWorkRequest(
+    worker: KClass<out DefaultWorker>,
+    userId: UserId? = null
+): OneTimeWorkRequest {
+    val inputData = WrapperWorkerFactory.workData(worker, userId)
+    val connectedConstraint = Constraints.Builder()
+        .setRequiredNetworkType(NetworkType.CONNECTED)
+        .build()
+    return OneTimeWorkRequest.Builder(workerClass)
+        .setConstraints(connectedConstraint)
+        .setExpedited(OutOfQuotaPolicy.RUN_AS_NON_EXPEDITED_WORK_REQUEST)
+        .setInputData(inputData)
+        .build()
+}
+
+private const val TIME_OF_EXECUTION = 4  // schedule at 4AM
```

Comment:
I wonder if we won't create a spike at 4AM on the backend because of this 😄

Although it is spread through different timezones, it could lead to some peaks. I won't advocate for over-engineering, but if we do need to fix it in the future, I guess we could schedule to be `lastExecutionDate + 24h` or something like that. So it will be spread across different users more randomly.

Reply:
It wouldn't be exactly 4AM for everyone, work manager uses some flex windows (I think it's like 15mins by default) but I added some randomizer to have some deviation (+/- 60mins so it gives a 2h window for each time zone) to spread it more evenly